Add support and testing for Python 3.12 and 3.13

pevogam opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Chores**
  * Expanded CI Python compatibility to include 3.12 and 3.13.
  * Centralized the Python version matrix for reuse across CI jobs.
  * Improved Python setup and enabled dependency caching to speed CI runs.
  * Added packaging dependencies for more reliable builds.
  * Restricted legacy egg builds to older Python versions (3.9–3.11); skipped for 3.12/3.13.
  * No functional/product behavior changes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pevogam committed:
Add support and testing for Python 3.12 and 3.13

Signed-off-by: Plamen Dimitrov <plamen.dimitrov@intra2net.com>

PaulYuuu commented:
Hello @pevogam, `distutils` was deprecated, if we plan to enable the CI check for python3.12 and python3.13, we should consider this.

I had a PR for distutils, but it has some risks, what's your opinion of the `distutils` module?

pevogam commented:
> `distutils`

Hi @PaulYuuu I think we should move forward from deprecated packages like this and comply with PEP 632.

Can you add a reference/link to the distutils pull request you proposed?

PaulYuuu commented:
> > `distutils`
> 
> Hi @PaulYuuu I think we should move forward from deprecated packages like this and comply with PEP 632.
> 
> Can you add a reference/link to the distutils pull request you proposed?

ah sorry, should be this one: https://github.com/avocado-framework/avocado-vt/pull/3920

pevogam commented:
Hi @PaulYuuu, the PR that we thought is a blocker for this is merged but I still get some problems here like
```
Run python setup.py bdist_egg
  python setup.py bdist_egg
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.12.11/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.12.11/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.11/x64
    Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.11/x64
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.11/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.12.11/x64/lib
Traceback (most recent call last):
  File "/home/runner/work/avocado-vt/avocado-vt/setup.py", line 17, in <module>
    from setuptools import find_packages, setup
ModuleNotFoundError: No module named 'setuptools'
Error: Process completed with exit code 1.
```
So I assume we have to deprecate the setuptools as well.

PaulYuuu commented:
> Hi @PaulYuuu, the PR that we thought is a blocker for this is merged but I still get some problems here like
> 
> ```
> Run python setup.py bdist_egg
>   python setup.py bdist_egg
>   shell: /usr/bin/bash -e {0}
>   env:
>     pythonLocation: /opt/hostedtoolcache/Python/3.12.11/x64
>     PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.12.11/x64/lib/pkgconfig
>     Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.11/x64
>     Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.11/x64
>     Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.11/x64
>     LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.12.11/x64/lib
> Traceback (most recent call last):
>   File "/home/runner/work/avocado-vt/avocado-vt/setup.py", line 17, in <module>
>     from setuptools import find_packages, setup
> ModuleNotFoundError: No module named 'setuptools'
> Error: Process completed with exit code 1.
> ```
> 
> So I assume we have to deprecate the setuptools as well.

Hello @pevogam, Python >= 3.12 will not pre-install the setuptools module, see https://docs.python.org/3/whatsnew/3.12.html
But avocado-vt uses `setuptools` as the build backend system.
However `python setup.py` will not install required packages before building the wheel/egg. In the future when we switch to use `python -m build`, the issue should be gone.
```
uv run python -m build
* Creating isolated environment: venv+pip...
* Installing packages in isolated environment:
  - setuptools >= 40.8.0
* Getting build dependencies for sdist...
running egg_info
creating avocado_framework_plugin_vt.egg-info
writing avocado_framework_plugin_vt.egg-info/PKG-INFO
```

In short, to fix the setuptools problem in this PR, please consider adding `setuptools;python_version >= '3.12'` in the `requirements-travis.txt` file.

PaulYuuu commented:
@pevogam , sorry I was in the wrong context, `name: Build Package for Python ${{ matrix.python-version }}` will not install `requirements-travis.txt`, so have to updte its 
```
      run: |
        python -m pip install --upgrade pip
        pip install build
```
to
```
      run: |
        python -m pip install --upgrade pip
        pip install build setuptools
```

For CI env, I don't think it's a problem that install setuptools even if it's installed(python<=3.11). But let's listen the voice from @richtja and @luckyh

pevogam commented:
Everything passes now, @richtja, @luckyh, let us know if you can review these changes too.

pevogam committed:
Reuse the python version list via a YAML anchor

Signed-off-by: Plamen Dimitrov <plamen.dimitrov@intra2net.com>

pevogam committed:
Use better pip-cached and latest minor python version in the CI

Signed-off-by: Plamen Dimitrov <plamen.dimitrov@intra2net.com>

pevogam committed:
Make egg builds a legacy behavior in the CI

Limit the egg build to earlier python versions only.

Also add wheel as it is lightweight and can handle edge-case
failures.

Signed-off-by: Plamen Dimitrov <plamen.dimitrov@intra2net.com>

pevogam commented on .github/workflows/ci.yml:
```diff
@@ -35,7 +35,7 @@ jobs:
 
     strategy:
       matrix:
-        python-version: ["3.9", "3.10", "3.11"]
+        python-version: &py_versions ["3.9", "3.10", "3.11", "3.12", "3.13"]
       fail-fast: false
```
Wrong, this agent already looped through the same cycle or recommendations a few times.

pevogam commented:
Alright, @PaulYuuu I think now I made both you and the rabbit bot happy (the latter happy enough as it cannot be easily satiated) so let me know if the final commits and changes look good enough to you.

luckyh reviewed (approved):
@pevogam thank you for the enhancement!

LGTM

pevogam commented:
Hi @PaulYuuu, I just fetched and rebased but I don't see it being removed, can you point me to exact commit hash?

PaulYuuu commented:
> Hi @PaulYuuu, I just fetched and rebased but I don't see it being removed, can you point me to exact commit hash?

Should be b6ba9b7b2a4c6c60a96eb2f848be714434c3bcca, and you can check the file from web to view.

pevogam commented:
I see, the time the pull request added 3.12 and 3.13 was before the 3.8 was dropped on master so this change ended up reintroducing the 3.8. Nice catch, must be fixed now.

richtja reviewed (approved):
@pevogam LGTM, thank you

luckyh commented:
Merging this, thanks to all!
